src/metrics/representation.py

```diff
@@ -0,0 +1,388 @@
+                    ].
+                closest: List of numpy arrays, with nearest searched indexes by top k, with its distances.
+                    Example for k = 3:
+                    [
+                        np.array([[4.        , 0.29289323],
+                               [2.        , 0.29289323],
+                               [6.        , 0.42264974]]), 
+                        np.array([[5.        , 0.29289323],
+                               [2.        , 0.29289323],
+                               [6.        , 0.42264974]]), 
+                        np.array([[4.        , 0.29289323],
+                               [5.        , 0.29289323],
+                               [6.        , 0.42264974]])
+                    ].
+            """
+            for i in range(0, len(queries), search_batch_size):
+                query_idxs = np.arange(i, min(i + search_batch_size, len(queries)))
+
+                closest_dist, closest_idx = index.search(queries[query_idxs], k=k)
+                relevant = relevants[query_idxs]
+ 
+                if self.metric_distance == MetricDistance.IP:
+                    closest_dist = 1 - closest_dist
+            
+                closest = map(lambda idx: np.stack((db_ids[closest_idx[idx]], closest_dist[idx]), axis=1), \
+                    np.arange(len(closest_idx)))
+                
+                if scores is None:
+                    relevant = map(lambda r: np.stack((r, np.ones_like(r)), axis=1), relevant)
+                else:
+                    relevant = map(lambda r_q: np.stack((r_q[0], scores[r_q[0], q_order_idxs[r_q[1]]]), axis=1), \
+                        zip(relevant, query_idxs))
+                
+                relevant = list(relevant)
+                closest = list(closest)
+                yield relevant, closest
+
+        return generator()
+
+    def build_index(self, vectors: np.ndarray):
+        """Build index of a given set of vectors with FAISS.
+
+        Args:
+            vectors: Database vectors to index, size (database_size, embedding_size).
+
+        Returns:
+            Constructed index.
+        """
+        vectors = vectors.astype(np.float32)
+        n, d = vectors.shape
+
+        index_class = faiss.IndexFlatIP if self.metric_distance == MetricDistance.IP else faiss.IndexFlatL2
+        if self.exact_index:
+            index = index_class(d)
+        else:
+            nlist = 4 * math.ceil(n ** 0.5)
+            quantiser = index_class(d)
+            index = faiss.IndexIVFFlat(quantiser, d, nlist, self.metric_distance.value)
+            index.train(vectors)
+
+        index.add(vectors)
+        return index
+
+
+# @METRICS.register_class
+class PrecisionAtKMeter(IndexBasedMeter):
+    def compute(self):
+        scores = []
+        generator = super().compute()
+        for relevant_idx, closest_idx in generator:
+            scores += precision(relevant_idx, closest_idx, k=self.k).tolist()
```

Comment:
Since only the function is changed, you can create a callable `metric_func` and change it in the inherited classes